compiler/rustc_target/src/spec/wasm_base.rs

```diff
@@ -95,9 +94,6 @@ pub fn options() -> TargetOptions {
 
         pre_link_args,
 
-        // FIXME: Figure out cases in which WASM needs to link with a native toolchain.
-        link_self_contained: LinkSelfContainedDefault::True,
```

Comment:
wasm32-unknown-unknown is always fully self contained. Both for linker and libraries. Why did you change it to no longer allow `-Clink-self-contained`?

Reply:
It was previously said that a self contained build was only supported for the `wasi` environment [RELEASES.md](https://github.com/rust-lang/rust/blob/master/RELEASES.md#compiler-17), https://github.com/rust-lang/rust/pull/104137#discussion_r1020200709

Do we also support `-Clink-self-contained` for wasm32-unknown-emscripten?

Reply:
`echo "fn main() {}" | rustc -Clink-self-contained=yes --target wasm32-unknown-unknown -` works just fine. I don't think `-Clink-self-contained=yes` has any effect though.

Reply:
Shall I set `LinkSelfContainedDefault::True` in `wasm32_unknown_unknown.rs` then?

Reply:
I think so. Probably same for wasm64-unknown-unknown.

Reply:
So my point in https://github.com/rust-lang/rust/pull/104137#discussion_r1020200709 was to allow `-Clink-self-contained` only when it's 1) supported 2) has any effect.

The second case is not true for wasm32/64 targets and for many other bare metal targets too, I guess.
It's unlikely that this option is currently used in practice with targets where it has no effect.

In theory, when https://github.com/rust-lang/rust/pull/96884 is implemented, `-Clink-self-contained` should start having an effect for these targets by enabling/disabling the shipped LLD.
So I wanted to prohibit `-Clink-self-contained` for as many targets as possible sooner, as a future-proofing measure.

Reply:
Makes sense.

Reply:
Just as a heads up, it seems like `-Clink-self-contained` (especially its default value of "True" here) did have a meaningful (side-)effect for wasm32 (and presumably also wasm64): It caused the subsequent link step to include `-nostartfiles`, which we need on wasm32-unknown-unknown so that it will not attempt to link in `crt1.o` with other linkers like `clang`. See #108910 for more details.

Reply:
(also, @cuviper has [suggested on zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/wasm.20custom.20linker.20breakage.20.23108910/near/340444726) that perhaps this example motivates a separate per-target flag for controlling whether we have crt objects.)

Reply:
(marking as unresolved to make this conversation easier to see while #108910 is open and others are skimming this to find connection with that issue.)